src/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjection.java

```diff
@@ -0,0 +1,494 @@
+			doJsonScan = true;
+			getMoreConfidence = true;
+			doAuthBypass = true;
+		}
+	}
+
+	@Override
+	public void scan(HttpMessage msg, NameValuePair originalParam) {
+		//TODO add TYPE_JSON control as soon as available
+		isJsonPayload = originalParam.getType() == NameValuePair.TYPE_POST_DATA; //& originalParam.getType() == NameValuePair.TYPE_JSON;
+		super.scan(msg, originalParam);
+	}
+
+	@Override
+	public void scan(HttpMessage msg, String param, String value) {
+		boolean isBingo = false;
+		List<Integer> rtt = new ArrayList<>();
+		if(!inScope(getTech())){
+			return;
+		}
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("\nScannning URL ["+msg.getRequestHeader().getMethod()+"] ["+msg.getRequestHeader().getURI()+
+					"] on param: ["+param+"] with value: ["+value+"] for MongoDB Injection");
+		}
+		// url-encoded query injection
+		if(doAllDataScan) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("\nStarting with the url-encoded query injection package of attack rules:");
+			}
+			int index = 0;
+			for(String valueInj : ALL_DATA_VALUE_INJECTION) {
+				String paramInj = param+ALL_DATA_PARAM_INJECTION[index++];
+				if(isStop()) {
+					if (LOG.isDebugEnabled()) { LOG.debug("Stopping the scan due to a user request"); }
+					return;
+				}
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("\nTrying with the value: "+paramInj+valueInj);
+				}
+				try {
+					HttpMessage msgInj = getNewMsg();
+					setParameter(msgInj, paramInj, valueInj);
+					sendAndReceive(msgInj, false);
+					rtt.add(msgInj.getTimeElapsedMillis());
+					String textBase = getBaseMsg().getResponseBody().toString();
+					String textInj = msgInj.getResponseBody().toString();
+					if(!textBase.equals(textInj)) {
+						// Get more confidence
+						if(getMoreConfidence) {
+							HttpMessage secondMsg = getNewMsg();
+							setParameter(secondMsg, param+URL_TOKEN, valueInj);
+							sendAndReceive(secondMsg, false);
+							String secondText = secondMsg.getResponseBody().toString();
+							if(!textInj.equals(secondText) && !areDifferentForOnlyValue(secondText, textInj, URL_TOKEN.
+									substring(1, URL_TOKEN.length()-1), paramInj.substring(1, paramInj.length()-1))) {
+								bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_HIGH, getName(), getDescription(), null, param,
+										paramInj+valueInj, getExtraInfo(ALL_DATA_ATTACK) , getSolution(), msgInj);
+								isBingo = true;
+								break;
+							}
+						}
+						else {
+							bingo(Alert.RISK_HIGH, Alert.CONFIDENCE_MEDIUM, getName(), getDescription(), null, param,
+									paramInj+valueInj, getExtraInfo(ALL_DATA_ATTACK) , getSolution(), msgInj);
+							isBingo = true;
+							break;
+						}
+					}
+				}catch(IOException ex) {
+					if (LOG.isDebugEnabled()) {
+						LOG.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() +" when "+URI_EX_LOG);
+					}					
+					return;
+				}
+			}
+		}
+		// not-handled error injection
+		if(doCrashScan) {
+			if (LOG.isDebugEnabled()) {

```

Comment:
This part should maybe pre-check that the original response didn't contain one of the error strings. If it was already some sort of failure there's no point re-testing it as it may lead to a false positive.

Reply:
I created a new list that contains only patterns that are not present in the original message